src/learn/internal/learnservice.cpp

```diff
@@ -0,0 +1,199 @@
+/*
+ * SPDX-License-Identifier: GPL-3.0-only
+ * MuseScore-CLA-applies
+ *
+ * MuseScore
+ * Music Composition & Notation
+ *
+ * Copyright (C) 2021 MuseScore BVBA and others
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "learnservice.h"
+
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QBuffer>
+#include <QtConcurrent>
+
+#include "log.h"
+#include "learnerrors.h"
+
+using namespace mu::learn;
+using namespace mu::network;
+
+void LearnService::init()
+{
+    refreshPlaylists();
+}
+
+Playlist LearnService::startedPlaylist() const
+{
+    return m_startedPlaylist;
+}
+
+mu::async::Channel<Playlist> LearnService::startedPlaylistChanged() const
+{
+    return m_startedPlaylistChannel;
+}
+
+Playlist LearnService::advancedPlaylist() const
+{
+    return m_advancedPlaylist;
+}
+
+mu::async::Channel<Playlist> LearnService::advancedPlaylistChanged() const
+{
+    return m_advancedPlaylistChannel;
+}
+
+void LearnService::openVideo(const std::string& videoId) const
+{
+    QUrl videoUrl = configuration()->videoOpenUrl(videoId);
+    interactive()->openUrl(videoUrl.toString().toStdString());
+}
+
+void LearnService::refreshPlaylists()
+{
+    async::Channel<RetVal<Playlist> >* startedPlaylistFinishChannel = new async::Channel<RetVal<Playlist> >();
+    startedPlaylistFinishChannel->onReceive(this, [this](const RetVal<Playlist>& result) {
+        if (!result.ret) {
+            LOGW() << result.ret.toString();
+            return;
+        }
+
+        if (m_startedPlaylist == result.val) {
+            return;
+        }
+
+        m_startedPlaylist = result.val;
+        m_startedPlaylistChannel.send(m_startedPlaylist);
+    });
+
+    async::Channel<RetVal<Playlist> >* advancedPlaylistFinishChannel = new async::Channel<RetVal<Playlist> >();
```

Comment:
Better to use channels just by value, this should work.

Reply:
thanks, fixed